test/test_util/test_data.py

```diff
@@ -0,0 +1,233 @@
+        warnings.simplefilter("ignore")
+        X_sampled, y_sampled = subsample(
+            X, y,
+            random_state=random_state,
+            sample_size=sample_size,
+            is_classification=task in CLASSIFICATION_TASKS
+        )
+
+    # Check the types remain the same
+    def dtype(obj):
+        if isinstance(obj, list):
+            if isinstance(obj[0], list):
+                return type(obj[0][0])
+            else:
+                return type(obj[0])
+        elif isinstance(obj, pd.DataFrame):
+            return obj.dtypes
+        else:
+            return obj.dtype
+
+    if isinstance(X, pd.DataFrame):
+        assert list(dtype(X_sampled)) == list(dtype(X))
+    else:
+        print(X)
+        assert dtype(X_sampled) == dtype(X)
+
+    if isinstance(y, pd.DataFrame):
+        assert list(dtype(y_sampled)) == list(dtype(y))
+    else:
+        assert dtype(y_sampled) == dtype(y)
+
+    # Check the right amount of samples were taken
+    def size(obj):
+        if isinstance(obj, spmatrix):
+            return obj.shape[0] if obj.shape[0] > 1 else obj.shape[1]
+        else:
+            return len(obj)
+
+    # check the right amount of samples were taken
+    if sample_size < 1:
+        assert size(X_sampled) == int(sample_size * size(X))
+    else:
+        assert size(X_sampled) == sample_size
+
+
+@parametrize('X', [np.asarray([[0, 0, 1]] * 10)])
+@parametrize('dtype', supported_precision_reductions + [np.dtype('float32'), np.dtype('float64')])
+@parametrize('x_type', [np.ndarray, csr_matrix])
+def test_reduce_precision_correctly_reduces_precision(X, dtype, x_type):
+    X = X.astype(dtype)
+    if x_type == csr_matrix:
+        X = x_type(X)
+
+    X_reduced, precision = reduce_precision(X)
+
+    # Check the reduced precision is correctly returned
+    assert X_reduced.dtype == precision
+
+    # Check that it was reduce to the correct precision
+    assert precision == reduction_mapping[dtype]
+
+    # Check that X's shape was not modified in any way
+    assert X.shape == X_reduced.shape
+
+    # Check that the return type is the one as we passed in
+    assert type(X) == type(X_reduced)
+
+
+@parametrize('X', [np.asarray([0, 0, 1]) * 10])
+@parametrize('dtype', [np.int32, np.int64, np.complex128])
+def test_reduce_precision_with_unsupported_dtypes(X, dtype):
+    X = X.astype(dtype)
+    with pytest.raises(ValueError):
+        reduce_precision(X)
+
+
+@parametrize("X", [
+    np.asarray([[1] * 10] * 100000, dtype=np.float64)  # Make it big for reductions to take place
+])
+@parametrize("x_type", [csr_matrix, np.ndarray])
+@parametrize("dtype", supported_precision_reductions)
+@parametrize('y, is_classification', [
+    (np.asarray([1] * 100000), True),
+    (np.asarray([1.0] * 100000), False),
+])
+@parametrize('multiplier', [1, 5.2, 10])
+@parametrize('operations', [['precision'], ['subsample'], ['precision', 'subsample']])
+def test_reduce_dataset_reduces_size_and_precision(
```

Comment:
Is there any chance to test for the expected subsample values as before? Or would there be too many options?

Reply:
That is possible, I would probably make them as their own separate test. I forgot to reinclude it. This test is more of a validation of it operation but not of specific values.

I'll do this

Reply:
So this is sort of checked already in `test_subsample_validaty`, given a `sample_size`, i.e. `0.25` or `10`, we check that `25%` of the samples remain or that `10` samples remain (consistent with the sklearn `train_test_split` api of allowing percentages or absolute number of samples.`

However it probably doesn't hurt to have explicit figures.

Reply:
Done